File: apps/web/src/lib/safe-to-spend.ts
// "Safe to Spend" — forward-looking runway to the next payday.
//
// Answers one question: "how much can I spend today without wrecking the plan?"
// It is deliberately distinct from the Budgets feature (backward-looking per-category caps).
// This is pure + side-effect-free (except the localStorage load/save helpers) so it can be
// unit-tested and reasoned about in isolation.
//
// The headline idea is the per-account FLOOR: money sitting below an account's floor is never
// counted as spendable. A current account defaults to a £0 floor (spend down to empty). An
// arranged-overdraft account can be given a negative floor (e.g. -£1000 = a 0% buffer line), and
// that floor can flip on a CLIFF date (e.g. the day the 0% overdraft ends and interest kicks in).

import type { Account, Bill } from "@/types/api";

export type PaydayConfig =
  | { kind: "lastWorkingDay" }
  | { kind: "dayOfMonth"; day: number };

export interface Cliff {
  accountId: number;
  /** ISO date (YYYY-MM-DD). On/after this local date, the account's floor becomes newFloorCents. */
  dateIso: string;
  newFloorCents: number;
}

export interface SafeToSpendConfig {
  payday: PaydayConfig;
  /** accountId -> floor in cents (may be negative for overdraft accounts). Missing = £0 floor. */
  floorsCents: Record<number, number>;
  cliffs: Cliff[];
  /** Extra money to set aside before dividing (planned debt paydown, etc.). */
  ringfenceCents: number;
  /** False until the user has set this up at least once; the tile shows a setup prompt instead. */
  configured: boolean;
}

export const DEFAULT_CONFIG: SafeToSpendConfig = {
  payday: { kind: "lastWorkingDay" },
  floorsCents: {},
  cliffs: [],
  ringfenceCents: 0,
  configured: false,
};

const STORAGE_KEY = "tally.safeToSpend";

export function loadConfig(): SafeToSpendConfig {
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return DEFAULT_CONFIG;
    return { ...DEFAULT_CONFIG, ...(JSON.parse(raw) as Partial<SafeToSpendConfig>) };
  } catch {
    return DEFAULT_CONFIG;
  }
}

export function saveConfig(cfg: SafeToSpendConfig): void {
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(cfg));
  } catch {
    // localStorage unavailable (private mode etc.) — non-fatal, config just won't persist.
  }
}

// ---- date helpers (all local time) ------------------------------------------------------------

function startOfDay(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth(), d.getDate());
}

const DAY_MS = 86_400_000;

/** Last Mon–Fri of the given month. month is 0-based. */
function lastWorkingDayOfMonth(year: number, month: number): Date {
  const d = new Date(year, month + 1, 0); // last calendar day of the month
  while (d.getDay() === 0 || d.getDay() === 6) d.setDate(d.getDate() - 1);
  return startOfDay(d);
}

/** The next payday strictly after `now` (a payday that falls on today rolls to next month). */
export function nextPayday(cfg: SafeToSpendConfig, now: Date): Date {
  const today = startOfDay(now);
  const build = (year: number, month: number): Date => {
    if (cfg.payday.kind === "lastWorkingDay") return lastWorkingDayOfMonth(year, month);
    const lastDay = new Date(year, month + 1, 0).getDate();
    return startOfDay(new Date(year, month, Math.min(cfg.payday.day, lastDay)));
  };
  const thisMonth = build(today.getFullYear(), today.getMonth());
  if (thisMonth.getTime() > today.getTime()) return thisMonth;
  return build(today.getFullYear(), today.getMonth() + 1);
}

// ---- floor resolution -------------------------------------------------------------------------

/** The active floor (cents) for an account, applying the latest cliff that has already passed. */
export function floorFor(account: Account, cfg: SafeToSpendConfig, now: Date): number {
  let floor = cfg.floorsCents[account.id] ?? 0;
  const today = startOfDay(now).getTime();
  let bestCliff = -Infinity;
  for (const c of cfg.cliffs) {
    if (c.accountId !== account.id) continue;
    const t = startOfDay(new Date(c.dateIso + "T00:00:00")).getTime();
    if (t <= today && t > bestCliff) {
      bestCliff = t;
      floor = c.newFloorCents;
    }
  }
  return floor;
}

// ---- the calculation --------------------------------------------------------------------------

export interface PerAccountSpendable {
  accountId: number;
  name: string;
  currentCents: number;
  floorCents: number;
  spendableCents: number;
}

export interface SafeToSpendResult {
  safeTodayCents: number;
  safePerDayCents: number;
  spendableNowCents: number;
  committedCents: number;
  ringfenceCents: number;
  freeCents: number;
  daysLeft: number;
  nextPaydayUnix: number;
  perAccount: PerAccountSpendable[];
}

function accountLabel(a: Account): string {
  return a.custom_display_name ?? a.consent_nickname ?? a.display_name;
}

/** Committed direct-debit / bill outflows between now and payday, deduped by name. */
export function committedBeforePayday(bills: Bill[], paydayUnix: number, now: Date): number {
  const nowUnix = Math.floor(startOfDay(now).getTime() / 1000);
  const seen = new Set<string>();
  let total = 0;
  for (const b of bills) {
    if (b.next_expected_date == null) continue;
    if (b.next_expected_date < nowUnix || b.next_expected_date > paydayUnix) continue;
    const key = b.name.replace(/\s*\(DD\)$/, "").toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    total += b.expected_amount_max_cents > 0
      ? b.expected_amount_max_cents
      : b.expected_amount_min_cents;
  }
  return total;
}

export function computeSafeToSpend(
  accounts: Account[],
  bills: Bill[],
  todaySpentCents: number,
  cfg: SafeToSpendConfig,
  now: Date
): SafeToSpendResult {
  const payday = nextPayday(cfg, now);
  const nextPaydayUnix = Math.floor(payday.getTime() / 1000);
  const daysLeft = Math.max(
    1,
    Math.ceil((payday.getTime() - startOfDay(now).getTime()) / DAY_MS)
  );

  const perAccount: PerAccountSpendable[] = [];
  let spendableNowCents = 0;
  for (const a of accounts) {
    if (a.kind !== "account" || a.current_balance_cents == null) continue;
    const floorCents = floorFor(a, cfg, now);
    const spendable = Math.max(0, a.current_balance_cents - floorCents);
    spendableNowCents += spendable;
    perAccount.push({
      accountId: a.id,
      name: accountLabel(a),
      currentCents: a.current_balance_cents,
      floorCents,
      spendableCents: spendable,
    });
  }

  const committedCents = committedBeforePayday(bills, nextPaydayUnix, now);
  const freeCents = spendableNowCents - committedCents - cfg.ringfenceCents;
  const safePerDayCents = Math.floor(freeCents / daysLeft);
  const safeTodayCents = safePerDayCents - todaySpentCents;

  return {
    safeTodayCents,
    safePerDayCents,
    spendableNowCents,
    committedCents,
    ringfenceCents: cfg.ringfenceCents,
    freeCents,
    daysLeft,
    nextPaydayUnix,
    perAccount,
  };
}

// ---- "can I afford this?" ----------------------------------------------------------------------

export interface AffordResult {
  ok: boolean;
  newSafePerDayCents: number;
  /** How far the purchase pushes the chosen account below its floor (>0 = breach). */
  floorBreachCents: number;
  reason: string;
}

export function affordCheck(
  amountCents: number,
  fromAccountId: number | null,
  accounts: Account[],
  bills: Bill[],
  todaySpentCents: number,
  cfg: SafeToSpendConfig,
  now: Date
): AffordResult {
  const base = computeSafeToSpend(accounts, bills, todaySpentCents, cfg, now);
  const newFree = base.freeCents - amountCents;
  const newSafePerDayCents = Math.floor(newFree / base.daysLeft);

  let floorBreachCents = 0;
  let breachName = "";
  if (fromAccountId != null) {
    const a = accounts.find((x) => x.id === fromAccountId);
    if (a && a.current_balance_cents != null) {
      const floor = floorFor(a, cfg, now);
      const after = a.current_balance_cents - amountCents;
      if (after < floor) {
        floorBreachCents = floor - after;
        breachName = accountLabel(a);
      }
    }
  }

  const ok = newSafePerDayCents >= 0 && floorBreachCents <= 0;
  let reason: string;
  if (floorBreachCents > 0) {
    reason = `Breaches your ${breachName} floor by ${pounds(floorBreachCents)}.`;
  } else if (newSafePerDayCents < 0) {
    reason = `Would put you ${pounds(-newSafePerDayCents)}/day in the red until payday.`;
  } else {
    reason = `Leaves you ${pounds(newSafePerDayCents)}/day until payday.`;
  }
  return { ok, newSafePerDayCents, floorBreachCents, reason };
}

function pounds(cents: number): string {
  return new Intl.NumberFormat("en-GB", { style: "currency", currency: "GBP" }).format(
    cents / 100
  );
}

File: apps/api/src/routes/api/safe_to_spend.rs
//! "Safe to spend" — forward-looking runway to the next payday.
//!
//! Mirrors the frontend calc in `apps/web/src/lib/safe-to-spend.ts` so the dashboard tile, the
//! `get_safe_to_spend` MCP tool and the daily Telegram ping all agree. Config is persisted as a
//! JSON blob in the existing `settings` key/value table (no new migration).
//!
//! The headline idea is the per-account FLOOR: money below an account's floor is never counted as
//! spendable. A current account defaults to a £0 floor; an arranged-overdraft account can have a
//! negative floor (a 0% buffer line), and that floor can flip on a CLIFF date (e.g. the day a 0%
//! overdraft ends and interest starts).

use crate::models::Bill;
use crate::AppState;
use axum::extract::State;
use axum::http::StatusCode;
use axum::Json;
use chrono::{Datelike, Duration, NaiveDate, Utc, Weekday};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

const CONFIG_KEY: &str = "safe_to_spend_config";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum Payday {
    LastWorkingDay,
    DayOfMonth { day: u32 },
}
impl Default for Payday {
    fn default() -> Self {
        Payday::LastWorkingDay
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Cliff {
    pub account_id: i64,
    /// ISO date (YYYY-MM-DD). On/after this date, the account's floor becomes new_floor_cents.
    pub date_iso: String,
    pub new_floor_cents: i64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Config {
    #[serde(default)]
    pub payday: Payday,
    #[serde(default)]
    pub floors_cents: HashMap<i64, i64>,
    #[serde(default)]
    pub cliffs: Vec<Cliff>,
    #[serde(default)]
    pub ringfence_cents: i64,
    #[serde(default)]
    pub configured: bool,
}

// ---- date + floor helpers ---------------------------------------------------------------------

fn last_day_of_month(year: i32, month: u32) -> u32 {
    let (ny, nm) = if month == 12 { (year + 1, 1) } else { (year, month + 1) };
    (NaiveDate::from_ymd_opt(ny, nm, 1).unwrap() - Duration::days(1)).day()
}

/// Last Mon–Fri of the given month.
fn last_working_day(year: i32, month: u32) -> NaiveDate {
    let last = last_day_of_month(year, month);
    let mut d = NaiveDate::from_ymd_opt(year, month, last).unwrap();
    while matches!(d.weekday(), Weekday::Sat | Weekday::Sun) {
        d -= Duration::days(1);
    }
    d
}

/// The next payday strictly after `today` (a payday falling on today rolls to next month).
fn next_payday(cfg: &Config, today: NaiveDate) -> NaiveDate {
    let build = |year: i32, month: u32| -> NaiveDate {
        match &cfg.payday {
            Payday::LastWorkingDay => last_working_day(year, month),
            Payday::DayOfMonth { day } => {
                let clamped = (*day).min(last_day_of_month(year, month)).max(1);
                NaiveDate::from_ymd_opt(year, month, clamped).unwrap()
            }
        }
    };
    let this = build(today.year(), today.month());
    if this > today {
        return this;
    }
    let (ny, nm) = if today.month() == 12 {
        (today.year() + 1, 1)
    } else {
        (today.year(), today.month() + 1)
    };
    build(ny, nm)
}

/// Active floor (cents) for an account, applying the latest cliff that has already passed.
fn floor_for(cfg: &Config, account_id: i64, today: NaiveDate) -> i64 {
    let mut floor = cfg.floors_cents.get(&account_id).copied().unwrap_or(0);
    let mut best: Option<NaiveDate> = None;
    for c in &cfg.cliffs {
        if c.account_id != account_id {
            continue;
        }
        if let Ok(d) = NaiveDate::parse_from_str(&c.date_iso, "%Y-%m-%d") {
            if d <= today && best.map_or(true, |b| d > b) {
                best = Some(d);
                floor = c.new_floor_cents;
            }
        }
    }
    floor
}

/// Committed direct-debit / bill outflows between `from_ts` and `payday_ts`, deduped by name
/// (stripping a trailing " (DD)" suffix, like the dashboard does).
fn committed_before_payday(bills: &[Bill], from_ts: i64, payday_ts: i64) -> i64 {
    let mut seen = HashSet::new();
    let mut total = 0i64;
    for b in bills {
        let Some(ned) = b.next_expected_date else { continue };
        if ned < from_ts || ned > payday_ts {
            continue;
        }
        let mut n = b.name.trim_end();
        if let Some(stripped) = n.strip_suffix("(DD)") {
            n = stripped.trim_end();
        }
        if !seen.insert(n.to_lowercase()) {
            continue;
        }
        total += if b.expected_amount_max_cents > 0 {
            b.expected_amount_max_cents
        } else {
            b.expected_amount_min_cents
        };
    }
    total
}

// ---- the calculation --------------------------------------------------------------------------

async fn load_config(state: &AppState) -> anyhow::Result<Config> {
    match state.db.get_setting(CONFIG_KEY).await? {
        Some(s) => Ok(serde_json::from_str(&s).unwrap_or_default()),
        None => Ok(Config::default()),
    }
}

/// Compute the safe-to-spend snapshot. Shared by the REST handler, the MCP tool and the scheduler.
/// Returns a camelCase JSON object matching the frontend `SafeToSpendResult`.
pub async fn compute(state: &AppState) -> anyhow::Result<Value> {
    let cfg = load_config(state).await?;
    let accounts = state.db.list_all_enabled_accounts().await?;
    let bills = state.db.list_bills().await?;

    let now = Utc::now();
    let today = now.date_naive();
    let now_ts = now.timestamp();
    let today_start = today.and_hms_opt(0, 0, 0).unwrap().and_utc().timestamp();

    let spend_rows = state.db.spending_by_category(today_start, now_ts).await?;
    let today_spent: i64 = spend_rows.iter().map(|(_, c)| *c).sum();

    let payday = next_payday(&cfg, today);
    let payday_ts = payday.and_hms_opt(23, 59, 59).unwrap().and_utc().timestamp();
    let days_left = (payday - today).num_days().max(1);

    let mut per_account = Vec::new();
    let mut spendable_now = 0i64;
    for a in &accounts {
        if a.kind != "account" {
            continue;
        }
        let Some(cur) = a.current_balance_cents else { continue };
        let floor = floor_for(&cfg, a.id, today);
        let spendable = (cur - floor).max(0);
        spendable_now += spendable;
        let name = a
            .custom_display_name
            .clone()
            .or_else(|| a.consent_nickname.clone())
            .unwrap_or_else(|| a.display_name.clone());
        per_account.push(json!({
            "accountId": a.id,
            "name": name,
            "currentCents": cur,
            "floorCents": floor,
            "spendableCents": spendable,
        }));
    }

    let committed = committed_before_payday(&bills, today_start, payday_ts);
    let free = spendable_now - committed - cfg.ringfence_cents;
    let safe_per_day = (free as f64 / days_left as f64).floor() as i64;
    let safe_today = safe_per_day - today_spent;

    Ok(json!({
        "safeTodayCents": safe_today,
        "safePerDayCents": safe_per_day,
        "spendableNowCents": spendable_now,
        "committedCents": committed,
        "ringfenceCents": cfg.ringfence_cents,
        "freeCents": free,
        "daysLeft": days_left,
        "spentTodayCents": today_spent,
        "nextPaydayUnix": payday.and_hms_opt(0, 0, 0).unwrap().and_utc().timestamp(),
        "configured": cfg.configured,
        "perAccount": per_account,
    }))
}

// ---- HTTP handlers ----------------------------------------------------------------------------

pub async fn get_status(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Value>, (StatusCode, String)> {
    Ok(Json(compute(&state).await.map_err(internal)?))
}

pub async fn get_config(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Value>, (StatusCode, String)> {
    let cfg = load_config(&state).await.map_err(internal)?;
    Ok(Json(serde_json::to_value(cfg).map_err(internal)?))
}

pub async fn put_config(
    State(state): State<Arc<AppState>>,
    Json(cfg): Json<Config>,
) -> Result<Json<Value>, (StatusCode, String)> {
    let s = serde_json::to_string(&cfg).map_err(internal)?;
    state.db.set_setting(CONFIG_KEY, &s).await.map_err(internal)?;
    Ok(Json(json!({ "ok": true })))
}

fn internal<E: std::fmt::Display>(e: E) -> (StatusCode, String) {
    (StatusCode::INTERNAL_SERVER_ERROR, format!("{e}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn cfg() -> Config {
        let mut floors = HashMap::new();
        floors.insert(1, 0); // Nationwide floor £0
        floors.insert(2, -100_000); // Santander floor -£1000
        Config {
            payday: Payday::LastWorkingDay,
            floors_cents: floors,
            cliffs: vec![Cliff {
                account_id: 2,
                date_iso: "2026-07-15".into(),
                new_floor_cents: 0,
            }],
            ringfence_cents: 0,
            configured: true,
        }
    }

    #[test]
    fn overdrawn_account_headroom_above_buffer() {
        // An account at -£832.32 against a -£1000 floor leaves £167.68 spendable — the
        // arranged overdraft counts as headroom only down to the floor, not to zero.
        let today = NaiveDate::from_ymd_opt(2026, 6, 20).unwrap();
        assert_eq!(floor_for(&cfg(), 2, today), -100_000);
        let spendable = (-83_232i64 - floor_for(&cfg(), 2, today)).max(0);
        assert_eq!(spendable, 16_768);
    }

    #[test]
    fn cliff_flips_floor_to_zero() {
        let before = NaiveDate::from_ymd_opt(2026, 7, 14).unwrap();
        let after = NaiveDate::from_ymd_opt(2026, 7, 15).unwrap();
        assert_eq!(floor_for(&cfg(), 2, before), -100_000);
        assert_eq!(floor_for(&cfg(), 2, after), 0);
    }

    #[test]
    fn payday_is_last_working_day_and_rolls_after() {
        // 20 Jun 2026 → next payday Tue 30 Jun (last working day of June).
        let today = NaiveDate::from_ymd_opt(2026, 6, 20).unwrap();
        let p = next_payday(&cfg(), today);
        assert_eq!(p, NaiveDate::from_ymd_opt(2026, 6, 30).unwrap());
        assert_eq!((p - today).num_days(), 10);
        // On payday itself it rolls to next month's last working day (Fri 31 Jul).
        let on_payday = NaiveDate::from_ymd_opt(2026, 6, 30).unwrap();
        assert_eq!(
            next_payday(&cfg(), on_payday),
            NaiveDate::from_ymd_opt(2026, 7, 31).unwrap()
        );
    }

    #[test]
    fn committed_dedups_and_windows() {
        let bills = vec![
            Bill {
                id: 1,
                name: "Energy".into(),
                expected_amount_min_cents: 11_000,
                expected_amount_max_cents: 11_700,
                currency: "GBP".into(),
                repeat_freq: "monthly".into(),
                next_expected_date: Some(
                    NaiveDate::from_ymd_opt(2026, 6, 26)
                        .unwrap()
                        .and_hms_opt(0, 0, 0)
                        .unwrap()
                        .and_utc()
                        .timestamp(),
                ),
                last_paid_date: None,
                match_description_regex: None,
                enabled: 1,
                source_recurring_id: None,
                created_at: 0,
                updated_at: 0,
            },
            // duplicate mandate name with " (DD)" suffix — must be deduped, not double-counted.
            Bill {
                id: 2,
                name: "Energy (DD)".into(),
                expected_amount_min_cents: 11_000,
                expected_amount_max_cents: 11_700,
                currency: "GBP".into(),
                repeat_freq: "monthly".into(),
                next_expected_date: Some(
                    NaiveDate::from_ymd_opt(2026, 6, 26)
                        .unwrap()
                        .and_hms_opt(0, 0, 0)
                        .unwrap()
                        .and_utc()
                        .timestamp(),
                ),
                last_paid_date: None,
                match_description_regex: None,
                enabled: 1,
                source_recurring_id: None,
                created_at: 0,
                updated_at: 0,
            },
        ];
        let from = NaiveDate::from_ymd_opt(2026, 6, 20)
            .unwrap()
            .and_hms_opt(0, 0, 0)
            .unwrap()
            .and_utc()
            .timestamp();
        let payday = NaiveDate::from_ymd_opt(2026, 6, 30)
            .unwrap()
            .and_hms_opt(23, 59, 59)
            .unwrap()
            .and_utc()
            .timestamp();
        assert_eq!(committed_before_payday(&bills, from, payday), 11_700);
    }
}
